File: Java/题目1150：Counterfeit Dollar.java

// 题目1150：Counterfeit Dollar

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:53
 * @url:http://ac.jobdu.com/problem.php?pid=1150
 * 解题思路参考csdn:http://blog.csdn.net/u013027996/article/details/17339141
 */

import java.util.Arrays;
import java.util.Scanner;

public class Main {
    /*
     * 1150
     */
    public static void main(String[] args) throws Exception {
        Scanner scanner = new Scanner(System.in);
        while (scanner.hasNext()) {
            int n = scanner.nextInt();
            //          ABCDEFGHIJKL
            for (int m = 0; m < n; m++) {
                int arrayA[] = new int[12];
                int arrayB[] = new int[12];
                Arrays.fill(arrayB, 1);
                int num = 0;
                for (int i = 0; i < 3; i++) {
                    String left = scanner.next();
                    String right = scanner.next();
                    String result = scanner.next();
                    int len1 = left.length();
                    int len2 = right.length();
                    if (result.equals("even")) {
                        for (int j = 0; j < len1; j++) {
                            arrayB[left.charAt(j) - 'A'] = 0;
                        }
                        for (int j = 0; j < len2; j++) {
                            arrayB[right.charAt(j) - 'A'] = 0;
                        }
                    } else if (result.equals("up")) {
                        num++;
                        for (int j = 0; j < len1; j++) {
                            arrayA[left.charAt(j) - 'A']--;
                        }
                        for (int j = 0; j < right.length(); j++) {
                            arrayA[right.charAt(j) - 'A']++;
                        }
                    } else if (result.equals("down")) {
                        num++;
                        for (int j = 0; j < left.length(); j++) {
                            arrayA[left.charAt(j) - 'A']++;
                        }
                        for (int j = 0; j < right.length(); j++) {
                            arrayA[right.charAt(j) - 'A']--;
                        }
                    }
                }

                int maxId = 0;
                int k = 0;
                while (k < 12) {
                    if (arrayB[k] == 1 && Math.abs(arrayA[k]) == num) {
                        maxId = k;
                    }
                    k++;
                }
                char c = (char) (maxId + 'A');
                if (arrayA[maxId] < 0) {
                    System.out.println(c
                            + " is the counterfeit coin and it is heavy.");
                } else {
                    System.out.println(c
                            + " is the counterfeit coin and it is light.");
                }
            }

        }
    }
}

/**************************************************************
 Problem: 1150
 User: wzqwsrf
 Language: Java
 Result: Accepted
 Time:80 ms
 Memory:15488 kb
 ****************************************************************/

File: C++/题目1150：Counterfeit Dollar.cpp

// 题目1150：Counterfeit Dollar

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:53
 * @url:http://ac.jobdu.com/problem.php?pid=1150
 * 解题思路参考csdn:http://blog.csdn.net/u013027996/article/details/17339141
 */

#include <stdio.h>
#include <string.h>

int arrayA[13];
int arrayB[13];
char array[13] = "ABCDEFGHIJKL";
int n, i, j, k;

int abs(int x) {
    return x < 0 ? -x : x;
}

void initArray() {
    for (i = 0; i < 13; i++) {
        arrayA[i] = 0;
        arrayB[i] = 1;
    }
}

int main() {
    while (scanf("%d", &n) != EOF) {
        char left[15], right[15], result[15];
        for (i = 0; i < n; i++) {
            int num = 0;
            initArray();
            for (j = 0; j < 3; j++) {
                scanf("%s %s %s", left, right, result);
                int len1 = strlen(left);
                int len2 = strlen(right);
                if (strcmp(result, "even") == 0) {
                    for (k = 0; k < len1; k++) {
                        arrayB[left[k] - 'A'] = 0;
                    }
                    for (k = 0; k < len2; k++) {
                        arrayB[right[k] - 'A'] = 0;
                    }
                } else if (strcmp(result, "up") == 0) {
                    num++;
                    for (k = 0; k < len1; k++) {
                        arrayA[left[k] - 'A']--;
                    }
                    for (k = 0; k < len2; k++) {
                        arrayA[right[k] - 'A']++;
                    }
                } else if (strcmp(result, "down") == 0) {
                    num++;
                    for (k = 0; k < len1; k++) {
                        arrayA[left[k] - 'A']++;
                    }
                    for (k = 0; k < len2; k++) {
                        arrayA[right[k] - 'A']--;
                    }
                }
            }
            int maxId = 0;
            for (j = 0; j < 13; j++) {
                if (arrayB[j] == 1 && abs(arrayA[j]) == num) {
                    maxId = j;
                }
            }
            char c = array[maxId];
            if (arrayA[maxId] < 0) {
                printf("%c is the counterfeit coin and it is heavy.\n", c);
            } else {
                printf("%c is the counterfeit coin and it is light.\n", c);
            }
        }
    }
    return 0;
}
/**************************************************************
Problem: 1150
User: wangzhenqing
Language: C++
Result: Accepted
Time:0 ms
Memory:1020 kb
****************************************************************/


                        

